src/main/java/com/patchworkmc/event/EventHandlerRewriter.java

```diff
@@ -66,35 +83,84 @@ public MethodVisitor visitMethod(int access, String name, String descriptor, Str
 		return new MethodScanner(parent, access, name, descriptor, signature);
 	}
 
+	/**
+	 * @deprecated glitch please clean this code up before merging it
+	 */
 	@Override
+	@Deprecated
 	public void visitEnd() {
-		// Write patchwork_registerStaticEventHandlers
-
-		if (staticSubscribers.isEmpty()) {
+		if (subscribeEvents.isEmpty()) {
 			super.visitEnd();
-
 			return;
 		}
 
-		Handle metafactory = new Handle(Opcodes.H_INVOKESTATIC, METAFACTORY_OWNER, "metafactory", METAFACTORY_DESCRIPTOR, false);
-		MethodVisitor staticRegistrar = super.visitMethod(Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC, REGISTER_STATIC, "(L" + EVENT_BUS + ";)V", null, null);
+		MethodVisitor staticRegistrar = null;
+		MethodVisitor instanceRegistrar = null;
 
-		if (staticRegistrar != null) {
+		if (hasStatic) {
+			staticRegistrar = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC, EventBus.REGISTER_STATIC, EventBus.REGISTER_STATIC_DESC, null, null);
 			staticRegistrar.visitCode();
+		}
+
+		if (hasInstance) {
+			instanceRegistrar = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC, EventBus.REGISTER_INSTANCE, EventBus.getRegisterInstanceDesc(className), null, null);
+			instanceRegistrar.visitCode();
+		}
+
+		// There are additional null checks in here to get IDEA to shut up about impossible NPEs
+		for (Map.Entry<SubscribeEvent, Boolean> subscribeEventBooleanEntry : subscribeEvents.entrySet()) {
+			SubscribeEvent subscriber = subscribeEventBooleanEntry.getKey();
+			boolean isStatic = subscribeEventBooleanEntry.getValue();
+
+			if (isStatic && staticRegistrar != null) {
+				Handle handler = new Handle(Opcodes.H_INVOKESTATIC, className, subscriber.getMethod(), subscriber.getMethodDescriptor(), false);
+
+				staticRegistrar.visitVarInsn(Opcodes.ALOAD, 0); // Load IEventBus on to the stack (1)
+				// Load the lambda on to the stack (2)
+				staticRegistrar.visitInvokeDynamicInsn("accept", "()Ljava/util/function/Consumer;", Lambdas.METAFACTORY, Lambdas.OBJECT_METHOD_TYPE, handler, Type.getMethodType(subscriber.getMethodDescriptor()));
+				// Pop eventbus and the lambda (0)
+				staticRegistrar.visitMethodInsn(Opcodes.INVOKEINTERFACE, EventBus.EVENT_BUS, "addListener", "(Ljava/util/function/Consumer;)V", true);
+			} else if (instanceRegistrar != null) {
+				instanceRegistrar.visitVarInsn(Opcodes.ALOAD, 1); // Load IEventBus on to the stack (1)
+				instanceRegistrar.visitVarInsn(Opcodes.ALOAD, 0); // Load <this> (technically we're in a static method but /shrug) (2)
+				instanceRegistrar.visitInsn(Opcodes.DUP); // Duplicate it (3)
+				// Ensure that it's not null (pop the duplicated <this>)
+				// side note: in java 8 this is a invokevirtual object.getClass(); pop;, but the J9+ version makes much more sense.
+				instanceRegistrar.visitMethodInsn(Opcodes.INVOKESTATIC, "java/util/Objects", "requireNonNull", "(Ljava/lang/Object;)", false);
```

Comment:
Why is this checking whether the instance is null inside the loop on every iteration? At the very least it should just check outside of the loop instead.

Reply:
it's doing that because Thats how JavaC Does It:tm:.
If it doesn't break fernflower i'll change it.

Reply:
I don't see why fernflower wouldn't handle it well, it's worth a test.

Reply:
it decompiles but with different behavior, but eh.